tools/purge_ems.rb

```diff
@@ -0,0 +1,125 @@
+  if ems.enabled?
+    log("Pausing management system with id: #{ems.id} #{ems.name}...")
+    ems.pause!
+    log("Pausing management system with id: #{ems.id} #{ems.name}...Complete")
+  end
+
+  log("Removing any workers for management system with id: #{ems.id} #{ems.name}...")
+  ems.ems_workers.each(&:kill_async)
+  ems.wait_for_ems_workers_removal
+  log("Removing any workers for management system with id: #{ems.id} #{ems.name}...Complete")
+end
+
+def existing_backlog?
+  # Watching the destroy queue messages be processed can take a long time so terminal disconnects
+  # are likely.  Therefore, if running it again, detect when it's in progress, skip queueing more
+  # and skip right to watching the updated progress if desired.
+  current_backlog > 0
+end
+
+def destroy_ems_in_batches(ems_id, batch, timeout, priority)
+  before_destroy_ems(ems_id)
+
+  if existing_backlog?
+    log("A backlog of work items is already in progress, skipping creating more.")
+    return
+  end
+
+  ems = ExtManagementSystem.find(ems_id)
+  log("Adding work items with batches of #{batch}...")
+  ems.class.reflections.select { |_, v| v.options[:dependent] }.map { |n, _v| ems.send(n) }.each do |rel|
+    next unless rel
+
+    log("  Adding #{rel.klass}")
+    rel.order(:id).pluck(:id).each_slice(batch) do |x|
+      MiqQueue.put(
+        :class_name  => rel.klass.to_s,
+        :method_name => DESTROY_METHOD,
+        :args        => [x],
+        :msg_timeout => timeout,
+        :priority    => priority,
+        :zone        => ZONE
+      )
+    end
+  end
+
+  log("Adding work items with batches of #{batch}...Complete")
+
+  MiqQueue.put(
+    :class_name  => ems.class,
+    :instance_id => ems.id,
+    :method_name => DESTROY_METHOD,
+    :msg_timeout => timeout,
+    :priority    => priority,
+    :zone        => ZONE
+  )
+end
+
+def destroy_parent_ems_in_batches(id, batch, timeout)
+  ems     = ExtManagementSystem.find(id)
+  timeout = timeout.to_i.minutes
+
+  log("Adding work items to remove child managers first...")
+  ems.child_manager_ids.each { |i| destroy_ems_in_batches(i, batch, timeout, PRIORITY - 10) }
+  log("Adding work items to remove child managers first...Complete")
+
+  log("Adding work items to remove targeted management system...")
+  destroy_ems_in_batches(ems.id, batch, timeout, PRIORITY)
+  log("Adding work items to remove targeted management system...Complete")
+end
+
+destroy_parent_ems_in_batches(*opts.values_at(:id, :batch, :timeout))
+
+if opts[:follow]
+  start_backlog = current_backlog
+  log("Destruction in progress...#{start_backlog} initial items")
+  require 'ruby-progressbar'
+
+  # Example format output:
+  # Progress: 5/10 |=====     |
+  pbar = ProgressBar.create(:title => "Progress", :total => start_backlog, :autofinish => false, :format => "%t: %c/%C |%B|")
+  loop do
+    remaining = current_backlog
+    pbar.progress = start_backlog - remaining
+
+    break if remaining == 0
+
+    sleep 10
```

Comment:
Incredibly minor, but we can break immediately after we get the remaining

```suggestion
    remaining = current_backlog
    break if remaining == 0

    pbar.progress = start_backlog - remaining
    sleep 10
```